Neos.Neos/Classes/Domain/Workspace/WorkspacePublisher.php

```diff
@@ -0,0 +1,220 @@
+<?php
+
+/*
+ * This file is part of the Neos.Neos package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+declare(strict_types=1);
+
+namespace Neos\Neos\Domain\Workspace;
+
+use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Command\DiscardIndividualNodesFromWorkspace;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Command\PublishIndividualNodesFromWorkspace;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Dto\NodeIdsToPublishOrDiscard;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Dto\NodeIdToPublishOrDiscard;
+use Neos\ContentRepository\Core\NodeType\NodeTypeName;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\FindClosestNodeFilter;
+use Neos\ContentRepository\Core\Projection\ContentGraph\VisibilityConstraints;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Annotations as Flow;
+use Neos\ContentRepository\Core\Feature\WorkspaceRebase\Command\RebaseWorkspace;
+use Neos\Neos\PendingChangesProjection\ChangeFinder;
+use Neos\Neos\PendingChangesProjection\ChangeProjection;
+
+/**
+ * Neos' workspace publisher service
+ *
+ * @api
+ */
+#[Flow\Scope('singleton')]
+readonly class WorkspacePublisher
+{
+    public function __construct(
+        private ContentRepositoryRegistry $contentRepositoryRegistry
+    ) {
+    }
```

Comment:
i would like to discuss whether we should have generally services scoped to one cr or the cr as id as argument.
Currently we follow the pattern of the `ContentRepositoryServiceFactoryInterface` which is also not pretty:
https://github.com/neos/neos-development-collection/issues/4634

but i guess having it uniform makes it easier to reason than say all neos.neos cr services are `singleton` with the `ContentRepositoryRegistry`, while core cr services or in framework agnostic code we are scoped by cr

Reply:
I'm not 100% sure where we draw the line what is a content repository service and what is not. In this case, I'd say this ist a higher level application service not integrated in the CR itself

Reply:
I would actually prefer not to use a cr service interface because of #4634 buut using the factory pattern and scoping services to a cr is a nice pattern which we use in other places as well, see the node creation stuff or catchup trigger things

```
interface NodeCreationHandlerFactoryInterface
{
    public function build(ContentRepository $contentRepository): NodeCreationHandlerInterface;
}
```

Reply:
As this a bit of a meta discussion (see https://github.com/neos/neos-development-collection/issues/4634) it should not block our efforts here, but it would be nice if we have a solution / pattern for this where we all agree upon.

_And_ deciding now would safe us needles refactorings lateron. I might as well be mistaken with my thesis to scope it to a cr, but i found it to be a clean pattern as this code is then technically easily portable as otherwise we would need to hardcode the `ContentRepositoryRegistry`. 

I would say it would be different if the `ContentRepositoryRegistry` would have an interface because that would make it a core pattern. But it was decided against that as far as i know.

Reply:
Additional note: One distinctive aspect would be that this adds no new functionality to the CR but only uses existing ones

Reply:
> In this case, I'd say this ist a higher level application service not integrated in the CR itself

I think the same, IMO it's fine to work with the CR registry at this level

Reply:
Okay two votes against one ;)